src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,393 @@
+
+    # 2. resolve function closure references
+    if func.__closure__ is not None:
+        closures = {
+            k: v
+            for k, v in zip(
+                code_object.co_freevars,
+                list(map(lambda x: x.cell_contents, func.__closure__)),
+            )
+        }
+        for k, v in closures.items():
+            ref_objects[k] = v
+            # if the closure item is a function and is not itself (recursive) and has not been visited
+            if isinstance(v, FunctionType) and k not in ref_objects and v != func:
+                get_func_references(v, ref_objects)
+
+
+def get_class_references(
+    cls: type, func_module_name: str, ref_objects: dict, classes_to_generate: list
+) -> None:
+    """
+    To get the referenced objects of a dynamically defined class.
+    A class could have methods, classes referencing objects.
+
+    """
+    # order matters to classes_to_generate, when constructing source code, referenced classes need to be defined first.
+    classes_to_generate.insert(0, cls)
+    # get base classes first
+    inferred_clses = [v for v in reversed(cls.__bases__) if v.__qualname__ != "object"]
+
+    # __dict__ contains function, classmethod, classes attributes within a given class which
+    # needs to be further analyzed
+    clsdict = dict(cls.__dict__)
+    for v in clsdict.values():
+        if isinstance(v, type) and v.__module__ == func_module_name:
+            # v is a class defined in the same module as UDF func's, need to dynamically parse the class
+            inferred_clses.append(v)
+        elif inspect.isfunction(v) or isinstance(v, classmethod):
+            # v is a function/classmethod, get the references objects of the Function object
+            get_func_references(
+                v if not isinstance(v, classmethod) else v.__func__, ref_objects
+            )
+
+    # recursively handling classes that should be parsed dynamically
+    for more_cls in inferred_clses:
+        get_class_references(
+            more_cls, func_module_name, ref_objects, classes_to_generate
+        )
+
+
+def extract_func_global_refs(code: CodeType) -> dict:
+    # inspired by cloudpickle to recursively extract all the global references used by the target func's code object
+    co_names = code.co_names
+    out_names = {}
+    for instr in dis.get_instructions(code):
+        op = instr.opcode
+        if op in GLOBAL_OPS:
+            out_names[co_names[instr.arg]] = None
+
+    if code.co_consts:
+        for const in code.co_consts:
+            if isinstance(const, CodeType):
+                out_names.update(extract_func_global_refs(const))
+
+    return out_names
+
+
+def remove_function_udf_annotation(
+    udf_source_code: str, code_as_comment: bool = True
+) -> str:
+    if code_as_comment:
+        return udf_source_code
+    # remove the udf/pandas_udf annotation to avoid re-registration
+    res = re.search(r"@(udf|pandas_udf)\(", udf_source_code)
```

Comment:
Should we assert that there is only one occurrence of such decorator? I understand that having two does not work anyway because you cannot execute `@udf` inside a UDF, but just curious if we want to throw this earlier.

Reply:
I honestly feel it's not the duty of the code generation -- it's more like udf source code validation.

and if 2 `@udf` is designed not to work, we probably should throw at an even earlier stage priori to source code generation.

Reply:
had some more thoughts on this after testing the current behavior.

`@udf` twice today would give us error:
`TypeError: <snowflake.snowpark.udf.UserDefinedFunction object at 0x121cff820> is not a module, class, method, or function.`

if code is generated as comments, we don't check and remove annotations, leaving the decorators as they're in the comments.

however I do feel checking multiple `@udf`s exist provide value for code being generated for execution (in the future).

I have updated the implementation to check and raise.

how do you feel now?